File: clients/js/src/features/config_resolver_ext.ts
import { Input } from "../input";

import {
  PolywrapClientConfigBuilder,
  PolywrapClient,
  Uri,
  ExtendableUriResolver
} from "@polywrap/client-js";
import {
  PluginPackage
} from "@polywrap/plugin-js";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    authority: unknown;
    result: unknown;
  }>(input);
  const authority = Input.expectString(inputObj.authority);
  const result = Input.expectString(inputObj.result);

  console.log("Creating CustomResolverExt Plugin");

  const customResolverExt = {
    uri: "wrap://plugin/custom-resolver",
    plugin: PluginPackage.from(() => ({
      tryResolveUri: async (args: { authority: string }) => {

        if (args.authority === authority) {
          return {
            manifest: null,
            uri: Uri.from(result).uri
          };
        }

        return null;
      }
    }))
  };

  console.log("Adding CustomResolverExt & ExtendableUriResolver to ClientConfig");

  const config = new PolywrapClientConfigBuilder()
    .setPackage(
      customResolverExt.uri,
      customResolverExt.plugin
    )
    .addInterfaceImplementation(
      ExtendableUriResolver.defaultExtInterfaceUris[0].uri,
      customResolverExt.uri
    )
    .addResolver(new ExtendableUriResolver())
    .build();

  const client = new PolywrapClient(config);

  console.log(`Resolving a wrap://${authority} URI`);

  const res = await client.tryResolveUri({
    uri: Uri.from(`wrap://${authority}/foo`)
  });

  if (res.ok && res.value.type === "uri") {
    console.log(`Received URI '${res.value.uri}'`);
    console.log("Success!");
  }
}

File: clients/py/src/features/config_resolver_ext.py
from typing import Any, TypedDict
from polywrap_core import Uri
from pydantic import BaseModel
import sys
from polywrap_client import PolywrapClient
from polywrap_uri_resolvers import ExtendableUriResolver
from polywrap_client_config_builder import PolywrapClientConfigBuilder
from polywrap_plugin import PluginModule, PluginPackage

from validators import UriStr


class Args(TypedDict):
    authority: str


class Plugin(PluginModule[None]):
    def __init__(self, config: None, authority: str, result: UriStr):
        super().__init__(config)
        self.authority = authority
        self.result = result
        setattr(self, "tryResolveUri", self.try_resolve_uri)

    def try_resolve_uri(self, args: Args, *_: Any):
        if args["authority"] == self.authority:
            return {"manifest": None, "uri": self.result.uri}
        return None


class TestCaseInput(BaseModel):
    authority: str
    result: UriStr


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    print("Creating CustomResolverExt Plugin")

    plugin = Plugin(None, input_obj.authority, input_obj.result)

    custom_resolver_ext_uri = Uri.from_str("wrap://plugin/custom-resolver")
    custom_resolver_ext_plugin = PluginPackage(plugin, NotImplemented)

    print("Adding CustomResolverExt & ExtendableUriResolver to ClientConfig")

    config = (
        PolywrapClientConfigBuilder()
        .set_package(custom_resolver_ext_uri, custom_resolver_ext_plugin)
        .add_interface_implementations(
            ExtendableUriResolver.DEFAULT_EXT_INTERFACE_URIS[0],
            [custom_resolver_ext_uri],
        )
        .add_resolver(ExtendableUriResolver())
        .build()
    )

    client = PolywrapClient(config)

    print(f"Resolving a wrap://{input_obj.authority} URI")

    res = client.try_resolve_uri(Uri.from_str(f"wrap://{input_obj.authority}/foo"))

    match res:
        case Uri() as uri:
            print(f"Received URI '{uri}'")
            print("Success!")
        case _:
            print("Failed to resolve URI.", file=sys.stderr)
